kv_store/lww_kvs/kvs_ebs_server.cpp

```diff
@@ -885,77 +891,119 @@ int main(int argc, char* argv[]) {
-          communication::Key_Request req;
-          req.set_sender("server");
-
-          // add each key that is going to this particular node
-          for (auto set_iter = map_iter->second.begin(); set_iter != map_iter->second.end(); set_iter++) {
-            communication::Key_Request_Tuple* tp = req.add_tuple();
-            tp->set_key(*set_iter);
+      // for any remote nodes that should receive gossip, we make key
+      // requests
+      for (auto map_iter = node_map.begin(); map_iter != node_map.end(); map_iter++) {
+        // create key request
+        communication::Key_Request req;
+        req.set_sender("server");
+
+        // add each key that is going to this particular node
+        for (auto set_iter = map_iter->second.begin(); set_iter != map_iter->second.end(); set_iter++) {
+          communication::Key_Request_Tuple* tp = req.add_tuple();
+          tp->set_key(*set_iter);
+          tp->set_global_memory_replication(placement[*set_iter].global_memory_replication_);
+          tp->set_global_ebs_replication(placement[*set_iter].global_ebs_replication_);
+        }
+
+        // send the request to the node
+        string key_req;
+        req.SerializeToString(&key_req);
+        zmq_util::send_string(key_req, &key_address_requesters[map_iter->first.key_exchange_connect_addr_]);
+
+        // receive the request
+        string key_res = zmq_util::recv_string(&key_address_requesters[map_iter->first.key_exchange_connect_addr_]);
+        communication::Key_Response resp;
+        resp.ParseFromString(key_res);
+
+        // for each key, add the address of *every* (there can be multiple)
+        // worker thread on the other node that should receive this key
+        for (int i = 0; i < resp.tuple_size(); i++) {
+          for (int j = 0; j < resp.tuple(i).address_size(); j++) {
+            (*res)[resp.tuple(i).address(j).addr()].insert(resp.tuple(i).key());
           }
+        }
+      }
 
-          // send the request to the node
-          string key_req;
-          req.SerializeToString(&key_req);
-          zmq_util::send_string(key_req, &key_address_requesters[map_iter->first.key_exchange_connect_addr_]);
-
-          // receive the request
-          string key_res = zmq_util::recv_string(&key_address_requesters[map_iter->first.key_exchange_connect_addr_]);
-          communication::Key_Response resp;
-          resp.ParseFromString(key_res);
-
-          // for each key, add the address of *every* (there can be multiple)
-          // worker thread on the other node that should receive this key
-          for (int i = 0; i < resp.tuple_size(); i++) {
-            for (int j = 0; j < resp.tuple(i).address_size(); j++) {
-              (*res)[resp.tuple(i).address(j).addr()].insert(resp.tuple(i).key());
-            }
+      // check if there are key requests to send to proxy
+      // NOTE: this is very similar to the code above, the only difference being the proxy address 
```

Comment:
This is bad. It seems like we should be able to factor out large chunks of this instead of repeating the code.